src/layers/layer.cpp

```diff
@@ -582,50 +528,189 @@ void Layer::write_proto(lbann_data::Layer* proto) const {
+    if (&get_child_layer(i) == &l) {
+      return i;
+    }
+  }
+  LBANN_ERROR(
+    l.get_type()," layer \"",l.get_name(),"\" ",
+    "is not a child layer of ",
+    this->get_type()," layer \"",this->get_name(),"\"");
+  return get_num_children();
+}
 
-  for (size_t i=1u; i < list.size(); ++i) {
-    if (list[i]) layer_names += ", " + list[i]->get_name();
+std::string Layer::get_parent_names() const {
+  std::ostringstream ss;
+  for (int i=0; i<get_num_parents(); ++i) {
+    ss << (i>0 ? ", " : "") << get_parent_layer(i).get_name();
+  }
+  return ss.str();
+}
+
+std::string Layer::get_child_names() const {
+  std::ostringstream ss;
+  for (int i=0; i<get_num_children(); ++i) {
+    ss << (i>0 ? ", " : "") << get_child_layer(i).get_name();
+  }
+  return ss.str();
+}
+
+void Layer::add_parent_layer(ViewingLayerPtr l) {
+  const auto* l_ptr = l.lock().get();
+  if (l_ptr == nullptr || l_ptr == this) {
+    return;
+  }
+  for (int i=0; i<get_num_parents(); ++i) {
+    if (l_ptr == &get_parent_layer(i)) {
+      return;
+    }
+  }
+  m_parent_layers.emplace_back(std::move(l));
+}
+
+void Layer::add_child_layer(ViewingLayerPtr l) {
+  const auto* l_ptr = l.lock().get();
+  if (l_ptr == nullptr || l_ptr == this) {
+    return;
+  }
+  for (int i=0; i<get_num_children(); ++i) {
+    if (l_ptr == &get_child_layer(i)) {
+      return;
+    }
+  }
+  m_child_layers.emplace_back(std::move(l));
+}
+
+void Layer::replace_parent_layer(ViewingLayerPtr l, size_t index) {
+  if (l.expired()) {
+    LBANN_ERROR(
+      "attempted to replace parent ",index," of ",
+      get_type()," layer \"",get_name(),"\" ",
+      "with an invalid layer pointer");
+  }
+  if (index >= m_parent_layers.size()) {
+    LBANN_ERROR(
+      "attempted to replace parent ",index," of ",
+      get_type()," layer \"",get_name(),"\", ",
+      "which only has ",m_parent_layers.size()," parents");
+  }
+  m_parent_layers[index] = std::move(l);
+}
+
+void Layer::replace_child_layer(ViewingLayerPtr l, size_t index) {
+  if (l.expired()) {
+    LBANN_ERROR(
+      "attempted to replace child ",index," of ",
+      get_type()," layer \"",get_name(),"\" ",
+      "with an invalid layer pointer");
+  }
+  if (index >= m_child_layers.size()) {
+    LBANN_ERROR(
+      "attempted to replace child ",index," of ",
+      get_type()," layer \"",get_name(),"\", ",
+      "which only has ",m_child_layers.size()," children");
   }
-  return layer_names;
+  m_child_layers[index] = std::move(l);
 }
 
-void Layer::add_parent_layer(const Layer* parent) {
-  const auto parent_pos = std::find(m_parent_layers.begin(),
-                                    m_parent_layers.end(),
-                                    parent);
-  if (parent != nullptr
-      && parent != this
-      && parent_pos == m_parent_layers.end()) {
-    m_parent_layers.push_back(parent);
+ViewingLayerPtr Layer::get_parent_layer_pointer(size_t index) const {
+  if (index >= m_child_layers.size()) {
```

Comment:
@timmoon10 this should be m_parent_layers.size()

Reply:
Good catch.